### PR TITLE
[Maven Extension] Support Maven Parallel Builds

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
**Description:**

Support [Maven Parallel Builds](https://cwiki.apache.org/confluence/display/MAVEN/Parallel+builds+in+Maven+3).

**Existing Issue(s):**

* https://github.com/open-telemetry/opentelemetry-java-contrib/issues/160

**Testing:**

Test the following build with the modified opentelemetry-maven-extension
```
otel-cli server tui
git clone https://github.com/apache/maven
cd maven
mvn dependency:copy -Dartifact=io.opentelemetry.contrib:opentelemetry-maven-extension:1.9.0-alpha
cp target/dependency/opentelemetry-maven-extension-1.9.0-alpha.jar /tmp/
mvn clean -Dotel.traces.exporter=otlp -Dotel.exporter.otlp.endpoint=http://localhost:4317 -Dmaven.ext.class.path=/tmp/opentelemetry-maven-extension-1.9.0-alpha.jar -T3
```

**Documentation:**

No change needed

**Outstanding items:**

None

### CODE REVIEW

File: maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java [hunk 1]
```diff
@@ -134,7 +148,9 @@ public static MavenProjectKey fromMavenProject(@Nonnull MavenProject mavenProjec
 
     abstract String pluginArtifactId();
 
-    static MojoExecutionKey fromMojoExecution(MojoExecution mojoExecution) {
+    abstract MavenProjectKey projectKey();
```
Comment (kenfinnigan) [inline maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:151]: Where is `projectKey()` used?

File: maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java [hunk 2]
```diff
@@ -114,9 +126,11 @@ public Span removeSpan(MojoExecution mojoExecution) {
 
     abstract String artifactId();
 
-    public static MavenProjectKey fromMavenProject(@Nonnull MavenProject mavenProject) {
+    abstract String version();
```
Comment (kenfinnigan) [inline maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:129]: Where is `version()` used?

**Review by kenfinnigan — COMMENTED** [review]
Looks reasonable, but had a question about a couple of new abstract methods

File: maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java (hunk 1 above)
Comment (cyrille-leclerc) [reply maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:151]: It is used by the Map of spans by `mojoExecutionKey` of the SpanRegistry. It's needed when using Maven Parallel Builds as the same mojo goal can be executed simultaneously on multiple Maven sub projects.

See https://github.com/open-telemetry/opentelemetry-java-contrib/blob/v1.9.0/maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java#L23

File: maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java (hunk 2 above)
Comment (cyrille-leclerc) [reply maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:129]: It's a protection for the map of spans by MojoExecutionKeys. I imagine 2 mojo executions can't execute simultaneously  on the same project with the different version of the Maven plugin but I felt it's a safe guardrail.

**Review by kenfinnigan — APPROVED** [review]

**Comment by kenfinnigan** [comment]
@trask Is there a problem with the `CODEOWNERS` config as I expected to be one of the allowed approvers, which doesn't seem to be the case.

**Comment by trask** [comment]
> @trask Is there a problem with the `CODEOWNERS` config as I expected to be one of the allowed approvers, which doesn't seem to be the case.

ya, the problem is that:

> The people you choose as code owners must have write permissions for the repository

(https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)

I'm not sure what the best workaround is. For now I have given everyone in `CODEOWNERS` write permission to the repo, and I'll seek guidance in the otel community repo.

**Review by trask — APPROVED** [review]

File: maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java [hunk 3]
```diff
@@ -38,6 +42,7 @@ public void setRootSpan(Span rootSpan) {
   }
 
   public Span getSpan(MavenProject mavenProject) {
+    logger.debug("OpenTelemetry: getSpan({}, {})", mavenProject, Thread.currentThread());
```
Comment (anuraaga) [inline maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:45]: I don't think we should have `Thread.currentThread` in logging messages - most loggers will output the thread info if configured in the pattern anyways.

If we do have it we to check if debug level is enabled before calling the not trivially cheap method

Comment (cyrille-leclerc) [reply maven-extension/src/main/java/io/opentelemetry/maven/SpanRegistry.java:45]: Thanks @anuraaga , fixed in PR https://github.com/open-telemetry/opentelemetry-java-contrib/pull/163

> most loggers will output the thread info if configured in the pattern anyways.

Unfortunately it's not the case with the `mvn --debug` / `mvn -X` option but we can remove the thread name as we understood the behaviour of Maven Parallel Builds.

**Pull request merged**